ompi/mca/op/rocm/op_rocm_functions.c

```diff
@@ -0,0 +1,1908 @@
+
+#include "ompi/mca/op/rocm/op_rocm.h"
+#include "ompi/mca/op/rocm/op_rocm_impl.h"
+
+/**
+ * Disable warning about empty macro var-args.
+ * We use varargs to suppress expansion of typenames
+ * (e.g., int32_t -> int) which could lead to collisions
+ * for similar base types. */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wgnu-zero-variadic-macro-arguments"
+
+static inline void device_op_pre(const void *orig_source1,
+                                 void **source1,
+                                 int *source1_device,
+                                 const void *orig_source2,
+                                 void **source2,
+                                 int *source2_device,
+                                 void *orig_target,
+                                 void **target,
+                                 int *target_device,
+                                 int count,
+                                 struct ompi_datatype_t *dtype,
+                                 int *threads_per_block,
+                                 int *max_blocks,
+                                 int *device,
+                                 opal_accelerator_stream_t *stream)
+{
+    uint64_t target_flags = -1, source1_flags = -1, source2_flags = -1;
+    int target_rc, source1_rc, source2_rc = -1;
+
+    *target = orig_target;
+    *source1 = (void*)orig_source1;
+    if (NULL != orig_source2) {
+        *source2 = (void*)orig_source2;
+    }
+
+    if (*device != MCA_ACCELERATOR_NO_DEVICE_ID) {
+        /* we got the device from the caller, just adjust the output parameters */
+        *target_device = *device;
+        *source1_device = *device;
+        if (NULL != source2_device) {
+            *source2_device = *device;
+        }
+    } else {
+
+        target_rc = opal_accelerator.check_addr(*target, target_device, &target_flags);
+        source1_rc = opal_accelerator.check_addr(*source1, source1_device, &source1_flags);
+        *device = *target_device;
+
+        if (NULL != orig_source2) {
+            source2_rc = opal_accelerator.check_addr(*source2, source2_device, &source2_flags);
+         }
+
+        if (0 == target_rc && 0 == source1_rc && 0 == source2_rc) {
+            /* no buffers are on any device, select device 0 */
+            *device = 0;
+        } else if (*target_device == -1) {
+            if (*source1_device == -1 && NULL != orig_source2) {
+                *device = *source2_device;
+            } else {
+                *device = *source1_device;
+            }
+        }
+
+        if (0 == target_rc || 0 == source1_rc || *target_device != *source1_device) {
+            size_t nbytes;
+            ompi_datatype_type_size(dtype, &nbytes);
+            nbytes *= count;
+
+            if (0 == target_rc) {
+                // allocate memory on the device for the target buffer
+                opal_accelerator.mem_alloc_stream(*device, target, nbytes, stream);
```

Comment:
not critiquing this particular stream allocation (or any subsequent ones),  I just would like to caution that the number of streams can have a performance impact overall, since it can influence how resources are mapped.